torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -1249,3 +1260,26 @@ def _cast_buffers_to_dtype_and_device(
             buffer.data = buffer.to(device=device)
         else:
             buffer.data = buffer.to(device=device, dtype=buffer_dtype)
+
+
+def _get_fsdp_states(module: nn.Module) -> List[_FSDPState]:
+    """
+    Returns all ``_FSDPState`` instances in the module tree rooted at
+    ``module`` without any duplicates and following the ``module.modules()``
+    traversal order.
+
+    For the wrapper code path, this returns all ``FullyShardedDataParallel``
+    instances. For the non-wrapper code path, this returns composable state
+    instances.
+
+    NOTE: For now, we must pass an ``nn.Module`` as the argument because
+    ``_FSDPState`` does not support graph traversal.
+    """
+    fsdp_states: List[_FSDPState] = []
+    visited_fsdp_states: Set[_FSDPState] = set()
+    for submodule in module.modules():
+        optional_state = _get_module_fsdp_state(submodule)
+        if optional_state is not None and optional_state not in visited_fsdp_states:
+            visited_fsdp_states.add(optional_state)
```

Comment:
do we need visited set if we just let fsdp_states be a set and convert back to a list?

Reply:
We need to preserve the ordering. Let me make that clear.